File: Backtracking/#46 Permutations.cpp
class Solution {
public:
    int n;
    vector<bool>vis;
    vector<int>res;
    vector<vector<int>> ans;
    void backtracking(vector<int> nums, int k){
        if(k==n){
            ans.push_back(res);
            return;
        }
        
        for(int i=0; i<n; i++){
            if(vis[i]) continue;
            res[k]=nums[i];
            vis[i]=true;
            backtracking(nums,k+1);
            vis[i]=false;
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        n = nums.size();
        vis = vector<bool>(n,0);
        res = vector<int>(n);
        backtracking(nums,0);
        return ans;
        
    }
};

File: Backtracking/#46. Permutations.py
'''
TC: O(N * N!)
SC: O(N!) for call stacks
'''
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.nums = nums
        self.gen_perm(nums, [])
        return self.res

    def gen_perm(self, design_space, temp):
        if not design_space:
            self.res.append(temp[:])
            return None

        for i in range(len(design_space)):
            temp.append(design_space[i])
            self.gen_perm(design_space[:i] + design_space[i+1:], temp)
            temp.pop()
        
